datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -219,42 +219,49 @@ fn quarter_month(date: &NaiveDateTime) -> u32 {
 /// account that some granularities are not uniform durations of time
 /// (e.g. months are not always the same lengths, leap seconds, etc)
 fn date_trunc_coarse(granularity: &str, value: i64) -> Result<i64> {
-    if granularity == "millisecond" || granularity == "microsecond" {
```

Comment:
this special case was what is confusing. I removed this one and inlined the `with_nanosecond(0)` call for all the other cases